### PR TITLE
feat(coderd/healthcheck): allow configuring database hc threshold

### LINKED ISSUE #9381: healthcheck: configurable database delay
The delay to mark the database as unhealthy should be configurable. currently, it looks like hardcoded to 15 milliseconds.

https://github.com/coder/coder/blob/b6e808d11628f6c407ca1c025a717edae0533173/coderd/healthcheck/database.go#L49-L51

I have a deployment where the remote db has 28-29 ms latency, and Coder is marking it as unhealthy.

image

### PR DESCRIPTION
**Pull request by johnstcn**
Fixes https://github.com/coder/coder/issues/9381:
- Exposes database healthcheck threshold in configuration

Also takes care of part of https://github.com/coder/coder/issues/8969:
- Exposes healthcheck recheck interval in configuration

### CODE REVIEW

File: cli/testdata/coder_server_--help.golden [hunk 1]
```diff
@@ -80,6 +80,18 @@ Use a YAML configuration file when your server launch become unwieldy.
           
           Write out the current server config as YAML to stdout.
 
+INTROSPECTION / HEALTHCHECK OPTIONS: 
+      --healthcheck-refresh duration, $CODER_HEALTHCHECK_REFRESH (default: 10m0s)
+          Refresh interval for healthchecks.
+
+      --threshold-database duration, $CODER_HEALTHCHECK_THRESHOLD_DATABASE (default: 15ms)
```
Comment (mafredri) [inline cli/testdata/coder_server_--help.golden:87]:
```suggestion
      --healthcheck-database-threshold duration, $CODER_HEALTHCHECK_THRESHOLD_DATABASE (default: 15ms)
```

Suggestion: unified prefix and reorder so future flags can be grouped together (--hc-db-query-latency, etc). Maybe we never add anything but it's nice to be prepared.

File: coderd/healthcheck/database.go [hunk 2]
```diff
@@ -10,20 +10,30 @@ import (
...
 type DatabaseReport struct {
-	Healthy   bool    `json:"healthy"`
-	Reachable bool    `json:"reachable"`
-	Latency   string  `json:"latency"`
-	LatencyMs int     `json:"latency_ms"`
-	Error     *string `json:"error"`
+	Healthy     bool    `json:"healthy"`
+	Reachable   bool    `json:"reachable"`
+	Latency     string  `json:"latency"`
+	LatencyMs   int64   `json:"latency_ms"`
```
Comment (mafredri) [inline coderd/healthcheck/database.go:22]: Do we typically style these as Ms or MS? I would think it's MS but I didn't verify, I don't really care which but would prefer consistency. Alt type it out, i.e. Millis or Milliseconds, but again, whichever is more consistent.

File: codersdk/deployment.go [hunk 3]
```diff
@@ -1799,6 +1812,37 @@ Write out the current server config as YAML to stdout.`,
 			Group:       &deploymentGroupClient,
 			YAML:        "webTerminalRenderer",
 		},
+		// Healthcheck Options
+		{
+			Name:        "Healthcheck Timeout",
```
Comment (mafredri) [inline codersdk/deployment.go:1817]:
```suggestion
			Name:        "Health Check Timeout",
```

I think it's ok in code, but as a compound word, healthcheck is uncommon. (Same for all other human readable texts.)

**Comment by matifali** [comment]
We should also show a hint on the database health check page on how to configure this interval. 
Probably expose it somewhere on the deployment page too. We are currently exposing many of the configured options there.

File: codersdk/deployment.go (hunk 3 above)
Comment (johnstcn) [reply codersdk/deployment.go]: Thoughts on `HEALTHCHECK` versus `HEALTH_CHECK`?

File: cli/testdata/coder_server_--help.golden (hunk 1 above)
Comment (johnstcn) [reply cli/testdata/coder_server_--help.golden]: This was an oversight. Fixed 👍

**Comment by johnstcn** [comment]
> We should also show a hint on the database health check page on how to configure this interval. Probably expose it somewhere on the deployment page too. We are currently exposing many of the configured options there.

Right now they'll show up by default under Deployment > Observability. I can move them to a separate option group and move them to a separate page if required.

I might need some design help with adding the hint to the health page though, so I'll defer that to a separate PR.

**Comment by matifali** [comment]
I think this is fine for now. A hint could be added in a separate PR. 

Thank you.

File: cli/testdata/coder_server_--help.golden [hunk 4]
```diff
@@ -80,6 +80,18 @@ Use a YAML configuration file when your server launch become unwieldy.
...
+      --health-check-refresh duration, $CODER_HEALTH_CHECK_REFRESH (default: 10m0s)
+          Refresh interval for healthchecks.
+
+      --health-check-threshold-database duration, $CODER_HEALTH_CHECK_THRESHOLD_DATABASE (default: 15ms)
+          The threshold for the database health check. If the median latency of
+          the database exceeds this threshold over 5 attempts, the database is
+          considered unhealthy. The default value is 15ms.
+
+      --health-check-timeout duration, $CODER_HEALTH_CHECK_TIMEOUT (default: 30s)
+          Overall timeout for health checks.
```
Comment (mtojek) [inline cli/testdata/coder_server_--help.golden:93]: nit: health check vs. healthcheck above

File: cli/testdata/server-config.yaml.golden [hunk 5]
```diff
@@ -232,6 +232,18 @@ introspection:
     # Allow administrators to enable Terraform debug output.
     # (default: false, type: bool)
     enableTerraformDebugMode: false
+  healthcheck:
+    # Overall timeout for health checks.
+    # (default: 30s, type: duration)
+    timeout: 30s
```
Comment (mtojek) [inline cli/testdata/server-config.yaml.golden:238]: Do we need to expose this configuration option? Wouldn't it be enough to keep it hardcoded for now?

File: coderd/healthcheck/database.go [hunk 6]
```diff
@@ -10,20 +10,30 @@ import (
...
-	Healthy   bool    `json:"healthy"`
-	Reachable bool    `json:"reachable"`
-	Latency   string  `json:"latency"`
-	LatencyMs int     `json:"latency_ms"`
-	Error     *string `json:"error"`
+	Healthy     bool    `json:"healthy"`
+	Reachable   bool    `json:"reachable"`
+	Latency     string  `json:"latency"`
+	LatencyMS   int64   `json:"latency_ms"`
+	ThresholdMS int64   `json:"threshold_ms"`
```
Comment (mtojek) [inline coderd/healthcheck/database.go:23]: Should we add the `Threshold string` for consistency? I don't know what is the original though behind string+int64 pairs.

File: cli/testdata/server-config.yaml.golden (hunk 5 above)
Comment (johnstcn) [reply cli/testdata/server-config.yaml.golden:238]: https://github.com/coder/coder/issues/8969 exists to expose recheck interval and I figured that the timeout would also be relevant. I don't have any strong opinions either way, so if you'd prefer it not exposed we can keep it hard-coded at 30s.

File: coderd/healthcheck/database.go (hunk 6 above)
Comment (johnstcn) [reply coderd/healthcheck/database.go:23]: Me either, and the fact that I don't know why this particular fence exists makes me pause before knocking it down. I do know that `_ms` was added post-hoc for the UI (8ee500c59d) so I suspect it was simply a case of not renaming a field in the API for compatibility purposes.

File: cli/testdata/server-config.yaml.golden (hunk 5 above)
Comment (mtojek) [reply cli/testdata/server-config.yaml.golden:238]: Is there any risk of misconfiguring it? Let's say an operator will set it to 5 mins.

Comment (johnstcn) [reply cli/testdata/server-config.yaml.golden]: OK, keeping timeout hard-coded for now. We can expose it if someone asks.

**Review by mtojek — APPROVED** [review]
This is in good shape to get merged 👍

File: cli/testdata/server-config.yaml.golden [hunk 7]
```diff
@@ -232,6 +232,15 @@ introspection:
...
     enableTerraformDebugMode: false
+  healthcheck:
+    # Refresh interval for healthchecks.
+    # (default: 10m0s, type: duration)
+    refresh: 10m0s
+    # The threshold for the database health check. If the median latency of the
+    # database exceeds this threshold over 5 attempts, the database is considered
+    # unhealthy. The default value is 15ms.
+    # (default: 15ms, type: duration)
+    thresholdDatabase: 15ms
```
Comment (mafredri) [inline cli/testdata/server-config.yaml.golden:243]: I think this would be smoother as `databaseThreshold` vs `thresholdDatabase` (same with flags, etc)

**Review by mafredri — APPROVED** [review]
Nice work!

**Pull request merged**